p2p/src/supervisor.rs

```diff
@@ -0,0 +1,517 @@
+
+        // FIXME(xla): No pipe should be of infinite length. Here the appropriate cap could be the
+        // maximal amount of inputs produced without another loop draining partially or fully.
+        let (input_tx, input_rx) = unbounded();
+
+        // FIXME(xla): The cap for this channel likely coincides with the maximum of connected
+        // peers.
+        let (accept_tx, accept_rx) = unbounded::<()>();
+        {
+            let input_tx = input_tx.clone();
+            let connected = connected.clone();
+            let name = "supervisor-accept".to_string();
+            let handle = thread::Builder::new()
+                .name(name.clone())
+                .spawn(|| Self::accept::<T>(accept_rx, connected, incoming, input_tx))?;
+
+            handles.push((name, handle));
+        }
+
+        // FIXME(xla): The cap for this channel likely coincides with the maximum of connected
+        // peers.
+        let (connect_tx, connect_rx) = unbounded::<transport::ConnectInfo>();
+        {
+            let input_tx = input_tx.clone();
+            let connected = connected.clone();
+            let name = "supervisor-connect".to_string();
+            let handle = thread::Builder::new()
+                .name(name.clone())
+                .spawn(|| Self::connect::<T>(connected, connect_rx, endpoint, input_tx))?;
+
+            handles.push((name, handle));
+        }
+
+        // FIXME(xla): Cap here amount of maximum peers by some arbitrary multiplier.
+        let (msg_tx, msg_rx) = unbounded::<(node::Id, message::Send)>();
+        {
+            let input_tx = input_tx.clone();
+            let peers = peers.clone();
+            let name = "supervisor-message".to_string();
+            let handle = thread::Builder::new()
+                .name(name.clone())
+                .spawn(move || Self::message::<T>(input_tx, msg_rx, peers))?;
+
+            handles.push((name, handle));
+        };
+
+        // FIXME(xla): Should be capped by maximum of connected peers.
+        let (stop_tx, stop_rx) = unbounded::<node::Id>();
+        {
+            let input_tx = input_tx.clone();
+            let peers = peers.clone();
+            let name = "supervisor-stop".to_string();
+            let handle = thread::Builder::new()
+                .name(name.clone())
+                .spawn(move || Self::stop::<T>(input_tx, peers, stop_rx))?;
+
+            handles.push((name, handle));
+        }
+
+        // FIXME(xla): Should be capped by maximum of connected peers.
+        let (upgrade_tx, upgrade_rx) = unbounded();
+        {
+            let peers = peers.clone();
+            let name = "supervisor-upgrade".to_string();
+            let handle = thread::Builder::new()
+                .name(name.clone())
+                .spawn(move || Self::upgrade::<T>(connected, input_tx, peers, upgrade_rx))?;
+
+            handles.push((name, handle));
+        }
+
+        let mut protocol = Protocol::default();
+
+        loop {
```

Comment:
Am I correct in saying that this is the core event loop of the `Supervisor`, and it handles message/event coordination across all worker threads and peers?

Also, similar to my earlier question, what happens if one of the worker threads die unexpectedly during this event loop's execution?

Reply:
> Am I correct in saying that this is the core event loop of the `Supervisor`, and it handles message/event coordination across all worker threads and peers (i.e. peer threads)?

Yes!



> Also, similar to my earlier question, what happens if one of the worker threads die unexpectedly during this event loop's execution?

See https://github.com/informalsystems/tendermint-rs/pull/708#discussion_r648137076